### PR TITLE
update 1.15.1

### PR DESCRIPTION
**Pull request by Arishamays1**
## ☆ Onnxruntime 1.15.1 Update ☆
[Jira Ticket](https://anaconda.atlassian.net/browse/PKG-1583)
[Upstream](https://github.com/microsoft/onnxruntime)
[Dependencies ](https://github.com/microsoft/onnxruntime/blob/main/requirements-dev.txt)
# Changes
- Updated version number and `sha256`
- Updated pinnings on dependencies
- A bit of reformatting of the `meta.yaml`
- removed `abs.yaml`
- Removed outdated patching, updated relevant ones
- Updated bld.bat and bld.sh for successful builds on archectures 
- `Onnxruntime` is available for python versions 3.7 - 3.10 so a `skip` has been added for versions greater than 10. See [here](https://github.com/microsoft/onnxruntime/blob/1b07bbceaa234a37d0418d8e61f2d52dbd80af37/setup.py#L507).

### CODE REVIEW

**Comment by sumit0190** [comment]
@Arishamays1 I took a look at the logs and made some changes that will hopefully get the build working:
- We can't build the package tests with our current setup. I faced this problem the last time too but unfortunately I couldn't find the time to locate and fix the problem. So the tests have been disabled and removed (I made the change only in `build.sh`, so a similar change has to be made in `bld.bat`).
- You need to call the `build` script with the `allow_running_as_root` option, as our CI workers do have root privileges. This is fine, and has no effect on the rest of the build process.
- You need to remove `libprotobuf` from the build environment, as that (and other `protobuf` related dependencies) are vendored in at the moment.
- You'll need to skip `s390x`, as big-endian machines are not suited for the ORT-formatted models used in this package.
- I haven't looked at the patches yet, so I am not sure if you need those.

File: recipe/bld.bat [hunk 1]
```diff
@@ -2,10 +2,11 @@
 
 :: We set CMAKE_DISABLE_FIND_PACKAGE_Protobuf=ON as currently we do not want to use
 :: protobuf from conda-forge, see https://github.com/conda-forge/onnxruntime-feedstock/issues/57#issuecomment-1518033552
-python tools/ci_build/build.py ^
+%python% tools/ci_build/build.py ^
+    --allow_running_as_root ^
```
Comment (sumit0190) [inline recipe/bld.bat]: You need to remove this option from `bld.bat`; it's for Unix only.

**Comment by sumit0190** [comment]
Seems like `ninja` was misbehaving a bit on `win-64`, so I removed it; hopefully `win-64` is able to build fully now (it did build correctly the last time, but the incorrect `makefile` behavior was causing `setup.py` to fail). Make sure to adjust the `meta.yaml` accordingly.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,73 +1,56 @@
...
     - {{ compiler('cxx') }}
     - cmake
     - git
-    - libprotobuf 3.20.1
-    - ninja
-    - patch    # [unix]
-    - m2-patch # [win]
+    - ninja     # [not win64]
     - pybind11
+    - numpy 1.21.5   # [py<310]
```
Comment (sumit0190) [inline recipe/meta.yaml]: Why are these in the `build` section AND the `host` section? Just putting them in `host` should be sufficient.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,73 +1,56 @@
...
     - packaging
-    - pip
-    - pybind11
-    - python
     - python-flatbuffers
-    - re2
-    - safeint
+    - optional-lite
+    - zlib
+    - numpy 1.21.5 # [py<310]
```
Comment (sumit0190) [inline recipe/meta.yaml]: Where is this version number (1.21.5) coming from? I don't see it in the upstream requirements, maybe this is not needed?

**Review by sumit0190 — APPROVED** [review]
LGTM apart from minor `numpy` related comments.

File: recipe/bld.bat [hunk 4]
```diff
@@ -1,42 +1,21 @@
...
-popd
-
-rem Create extra cmake defines and associated variables
-if %PKG_NAME% == "onnxruntime-novec" ( set DONT_VECTORIZE="ON" ) else set DONT_VECTORIZE="OFF"
-set CMAKE_EXTRA_DEFINES= "Protobuf_SRC_ROOT_FOLDER=%LIBRARY_PREFIX%\lib\pkgconfig" "onnxruntime_PREFER_SYSTEM_LIB=ON" "onnxruntime_USE_COREML=OFF" "onnxruntime_DONT_VECTORIZE=%DONT_VECTORIZE%" "onnxruntime_BUILD_SHARED_LIB=ON" "onnxruntime_BUILD_UNIT_TESTS=OFF" "CMAKE_PREFIX_PATH=%LIBRARY_PREFIX%" "CMAKE_CXX_FLAGS=/wd4100 /wd4244 /wd4127" "CMAKE_VERBOSE_MAKEFILE=ON" 
-
-%PYTHON% tools\ci_build\build.py --build_dir build-ci --config Release --update --build --skip_submodule_sync --build_wheel --parallel --enable_lto --use_full_protobuf --cmake_extra_defines %CMAKE_EXTRA_DEFINES%
+:: We set CMAKE_DISABLE_FIND_PACKAGE_Protobuf=ON as currently we do not want to use
+:: protobuf from conda-forge, see https://github.com/conda-forge/onnxruntime-feedstock/issues/57#issuecomment-1518033552
+%python% tools/ci_build/build.py ^
```
Comment (skupr-anaconda) [inline recipe/bld.bat]:
```suggestion
%PYTHON% tools/ci_build/build.py ^
```
Just a convention and for visibility

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,73 +1,56 @@
...
-    sha256: 5daca6ca216495edf89d167f808d1d03c4a4d929cef7da5e10f135ae1540c7e4
-    folder: json
-  - url: https://github.com/microsoft/wil/archive/e8c599bca6c56c44b6730ad93f6abbc9ecd60fc1.tar.gz  # [win]
-    sha256: 388156309853aaa6e01d177f9b68bd2c6943e5fc073ac6786f9383c1b4438812 # [win]
-    folder: wil # [win]
+      - never_call_install_python_deps.patch
+    #   Workaround for https://github.com/conda-forge/onnxruntime-feedstock/pull/56#issuecomment-1586080419
+      - windows_workaround_conflict_onnxruntime_dll_system32.patch  # [win]
+      - 16337.patch
+    #   Workaround for https://github.com/microsoft/onnxruntime/issues/13225
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]: Could you place a comment above the corresponding patch to avoid confusing?

File: recipe/meta.yaml [hunk 6]
```diff
@@ -1,73 +1,56 @@
...
-    - pip
-    - pybind11
-    - python
     - python-flatbuffers
-    - re2
-    - safeint
+    - optional-lite
+    - zlib
+    - numpy 1.21.5 # [py<310]
+    - numpy # [py>=310]
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]: Should it be like:
`- numpy {{ numpy }}  # [py>=310]` so, in runtime, a correct version will be picked up

File: recipe/meta.yaml [hunk 7]
```diff
@@ -1,73 +1,56 @@
...
+      - never_call_install_python_deps.patch
+    #   Workaround for https://github.com/conda-forge/onnxruntime-feedstock/pull/56#issuecomment-1586080419
+      - windows_workaround_conflict_onnxruntime_dll_system32.patch  # [win]
+      - 16337.patch
+    #   Workaround for https://github.com/microsoft/onnxruntime/issues/13225
 
 build:
   number: 0
-  skip: True # [s390x or ppc64le]
+  skip: True # [py>310 or ppc64le or s390x]
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]: It's outdated info that it does not support python 3.11 https://github.com/microsoft/onnxruntime/blob/1b07bbceaa234a37d0418d8e61f2d52dbd80af37/setup.py#L507. The support was added in v1.15.0 https://github.com/microsoft/onnxruntime/releases/tag/v1.15.0

File: recipe/meta.yaml [hunk 8]
```diff
@@ -1,73 +1,56 @@
...
   build:
+    - python
     - {{ compiler('c') }}
     - {{ compiler('cxx') }}
     - cmake
     - git
-    - libprotobuf 3.20.1
-    - ninja
-    - patch    # [unix]
-    - m2-patch # [win]
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]: They are required if we apply patches.
```
    - patch    # [unix]
    - m2-patch # [win]
```
And check the anaconda-linter error about missing them

File: recipe/meta.yaml [hunk 9]
```diff
@@ -1,73 +1,56 @@
...
+  missing_dso_whitelist:     # [win64]
+    - '**/api-ms-win-core-path-l1-1-0.dll'     # [win64]
 
 requirements:
   build:
+    - python
     - {{ compiler('c') }}
     - {{ compiler('cxx') }}
     - cmake
     - git
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]:
```suggestion
    - git  # [not win]
```
**Prefect** already has `git` on `win64`. Also, this change will prevent crashing the `win64` agent on Prefect. See a corresponding error in the anaconda-linter log message

**Comment by skupr-anaconda** [comment]
An error on `osx-64`:
```
dyld[1064]: Library not loaded: @rpath/libzstd.1.dylib

  Referenced from: <72742B14-65A8-3066-A0E4-314A22D5ECE8> /private/var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_ddtlf7r4qf/pkg_dirs/cmake-3.26.4-hbb7d193_0/bin/cmake

  Reason: tried: '/private/var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_ddtlf7r4qf/pkg_dirs/cmake-3.26.4-hbb7d193_0/bin/../lib/libzstd.1.dylib' (no such file), '/private/var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_ddtlf7r4qf/pkg_dirs/cmake-3.26.4-hbb7d193_0/bin/../lib/libzstd.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS@rpath/libzstd.1.dylib' (no such file), '/private/var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_ddtlf7r4qf/pkg_dirs/cmake-3.26.4-hbb7d193_0/bin/../lib/libzstd.1.dylib' (no such file), '/private/var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_ddtlf7r4qf/pkg_dirs/cmake-3.26.4-hbb7d193_0/bin/../lib/libzstd.1.dylib' (no such file), '/usr/local/lib/libzstd.1.dylib' (no such file), '/usr/lib/libzstd.1.dylib' (no such file, not in dyld cache)
```

**Comment by akabanovs** [comment]
> An error on `osx-64`:
> 
> ```
> dyld[1064]: Library not loaded: @rpath/libzstd.1.dylib
> 
>   Referenced from: <72742B14-65A8-3066-A0E4-314A22D5ECE8> /private/var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_ddtlf7r4qf/pkg_dirs/cmake-3.26.4-hbb7d193_0/bin/cmake
> 
>   Reason: tried: '/private/var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_ddtlf7r4qf/pkg_dirs/cmake-3.26.4-hbb7d193_0/bin/../lib/libzstd.1.dylib' (no such file), '/private/var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_ddtlf7r4qf/pkg_dirs/cmake-3.26.4-hbb7d193_0/bin/../lib/libzstd.1.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS@rpath/libzstd.1.dylib' (no such file), '/private/var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_ddtlf7r4qf/pkg_dirs/cmake-3.26.4-hbb7d193_0/bin/../lib/libzstd.1.dylib' (no such file), '/private/var/folders/c_/qfmhj66j0tn016nkx_th4hxm0000gp/T/abs_ddtlf7r4qf/pkg_dirs/cmake-3.26.4-hbb7d193_0/bin/../lib/libzstd.1.dylib' (no such file), '/usr/local/lib/libzstd.1.dylib' (no such file), '/usr/lib/libzstd.1.dylib' (no such file, not in dyld cache)
> ```

Perhaps we need zstd package? It seems osx-64 was previously built succesfully? Perhaps something else provided zstd?

File: recipe/meta.yaml (hunk 8 above)
Comment (Arishamays1) [reply recipe/meta.yaml]: Great! added this back. Thanks

File: recipe/meta.yaml (hunk 7 above)
Comment (Arishamays1) [reply recipe/meta.yaml]: Thank you for letting me know! The setup.py and the 1.15.1 tag were a bit misleading.

File: recipe/meta.yaml (hunk 3 above)
Comment (Arishamays1) [reply recipe/meta.yaml]: You're right! I believe this came from when python 3.11 was still unsupported for onnxruntime. Thanks!

File: recipe/meta.yaml (hunk 6 above)
Comment (Arishamays1) [reply recipe/meta.yaml]: Added, however, will the selector be necessary if python 3.11 is now fully supported? I went ahead and removed them

File: recipe/meta.yaml (hunk 2 above)
Comment (Arishamays1) [reply recipe/meta.yaml]: Thanks will do

File: recipe/meta.yaml (hunk 5 above)
Comment (Arishamays1) [reply recipe/meta.yaml]: done!

File: recipe/meta.yaml [hunk 10]
```diff
@@ -1,73 +1,54 @@
...
+    - optional-lite
+    - zlib
+    - zstd {{ zstd }}
+    - numpy {{ numpy }}
+    - pybind11
     - setuptools
-    - sympy >=1.10
     - wheel
   run:
     - coloredlogs
```
Comment (ryanskeith) [inline recipe/meta.yaml:54]: I think you are missing some dependencies.  https://github.com/microsoft/onnxruntime/blob/main/requirements-training.txt.  There are several requirements files.  I think the runtime one is the _training IF I am reading things correctly.

**Review by ryanskeith — CHANGES_REQUESTED** [review]
I think there are missing dependencies.  Also, I couldn't get the logs to download but there were some things on Prefect that had me worried that it wasn't finding libraries to build.  I think this library may also merit doing a full test to make sure it is functioning properly.

File: recipe/meta.yaml [hunk 11]
```diff
@@ -1,73 +1,54 @@
...
   number: 0
-  skip: True # [s390x or ppc64le]
+  skip: True # [ppc64le or s390x]
   ignore_run_exports_from:
-    - libprotobuf
+    - zlib
+  entry_points:
+    - onnxruntime_test = onnxruntime.tools.onnxruntime_test:main
+  missing_dso_whitelist:     # [win64]
+    - '**/api-ms-win-core-path-l1-1-0.dll'     # [win64]
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]:
```suggestion
  missing_dso_whitelist:     # [win]
    - '**/api-ms-win-core-path-l1-1-0.dll'     # [win]
```

File: recipe/meta.yaml [hunk 12]
```diff
@@ -1,73 +1,54 @@
...
   build:
+    - python
     - {{ compiler('c') }}
     - {{ compiler('cxx') }}
     - cmake
-    - git
-    - libprotobuf 3.20.1
-    - ninja
+    - git  # [not win]
+    - ninja     # [not win64]
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]:
```suggestion
    - git  # [not win]
    - ninja     # [not win]
```

**Review by skupr-anaconda — APPROVED** [review]
LGTM but `win` selector

**Pull request merged**